File: algo-ds-library/python/graphs/graph_traversal.py
"""
Graph Traversal: DFS & BFS
---------------------------
Graphs are represented as adjacency lists:
    graph = {
        'A': ['B', 'C'],
        'B': ['A', 'D'],
        ...
    }

DFS (Depth-First Search)
------------------------
Time Complexity:  O(V + E)  — V = vertices, E = edges
Space Complexity: O(V)      — visited set + recursion stack

BFS (Breadth-First Search)
--------------------------
Time Complexity:  O(V + E)
Space Complexity: O(V)      — visited set + queue
"""

from collections import deque


# ─── DFS ────────────────────────────────────────────────────────────────────

def dfs_recursive(graph: dict, start: str, visited: set = None) -> list:
    """
    Depth-First Search using recursion.
    Explores as deep as possible before backtracking.
    Returns list of nodes in visit order.
    """
    if visited is None:
        visited = set()

    visited.add(start)
    result = [start]

    for neighbour in graph.get(start, []):
        if neighbour not in visited:
            result.extend(dfs_recursive(graph, neighbour, visited))

    return result


def dfs_iterative(graph: dict, start: str) -> list:
    """
    Depth-First Search using an explicit stack.
    Equivalent traversal to recursive DFS.
    Returns list of nodes in visit order.
    """
    visited = set()
    stack = [start]
    result = []

    while stack:
        node = stack.pop()
        if node not in visited:
            visited.add(node)
            result.append(node)
            # Push neighbours in reverse so left-to-right order is preserved
            for neighbour in reversed(graph.get(node, [])):
                if neighbour not in visited:
                    stack.append(neighbour)

    return result


# ─── BFS ────────────────────────────────────────────────────────────────────

def bfs(graph: dict, start: str) -> list:
    """
    Breadth-First Search using a queue.
    Explores all neighbours at the current depth before going deeper.
    Returns list of nodes in visit order.
    """
    visited = {start}
    queue = deque([start])
    result = []

    while queue:
        node = queue.popleft()
        result.append(node)

        for neighbour in graph.get(node, []):
            if neighbour not in visited:
                visited.add(neighbour)
                queue.append(neighbour)

    return result


def bfs_shortest_path(graph: dict, start: str, end: str) -> list | None:
    """
    BFS to find the shortest path between start and end.
    Returns the path as a list of nodes, or None if no path exists.
    """
    if start == end:
        return [start]

    visited = {start}
    queue = deque([[start]])

    while queue:
        path = queue.popleft()
        node = path[-1]

        for neighbour in graph.get(node, []):
            if neighbour not in visited:
                new_path = path + [neighbour]
                if neighbour == end:
                    return new_path
                visited.add(neighbour)
                queue.append(new_path)

    return None  # No path found

File: algo-ds-library/java/graphs/GraphTraversal.java
package graphs;

import java.util.*;

/**
 * Graph Traversal: DFS & BFS
 *
 * Graph represented as an adjacency list.
 *
 * DFS (Depth-First Search)
 *   Time Complexity:  O(V + E)
 *   Space Complexity: O(V) — visited set + stack
 *
 * BFS (Breadth-First Search)
 *   Time Complexity:  O(V + E)
 *   Space Complexity: O(V) — visited set + queue
 */
public class GraphTraversal {

    private final Map<String, List<String>> adjacencyList;

    public GraphTraversal() {
        this.adjacencyList = new LinkedHashMap<>();
    }

    /** Add an undirected edge between two nodes. */
    public void addEdge(String a, String b) {
        adjacencyList.computeIfAbsent(a, k -> new ArrayList<>()).add(b);
        adjacencyList.computeIfAbsent(b, k -> new ArrayList<>()).add(a);
    }

    /** Add a node with no edges. */
    public void addNode(String node) {
        adjacencyList.putIfAbsent(node, new ArrayList<>());
    }

    // ── DFS Recursive ────────────────────────────────────────────────────────

    /**
     * DFS using recursion.
     * Explores as deep as possible before backtracking.
     * Returns nodes in visit order.
     */
    public List<String> dfsRecursive(String start) {
        List<String> result = new ArrayList<>();
        Set<String> visited = new HashSet<>();
        dfsHelper(start, visited, result);
        return result;
    }

    private void dfsHelper(String node, Set<String> visited, List<String> result) {
        visited.add(node);
        result.add(node);
        for (String neighbour : adjacencyList.getOrDefault(node, Collections.emptyList())) {
            if (!visited.contains(neighbour)) {
                dfsHelper(neighbour, visited, result);
            }
        }
    }

    // ── DFS Iterative ────────────────────────────────────────────────────────

    /**
     * DFS using an explicit stack.
     * Returns nodes in visit order.
     */
    public List<String> dfsIterative(String start) {
        List<String> result = new ArrayList<>();
        Set<String> visited = new HashSet<>();
        Deque<String> stack = new ArrayDeque<>();
        stack.push(start);

        while (!stack.isEmpty()) {
            String node = stack.pop();
            if (!visited.contains(node)) {
                visited.add(node);
                result.add(node);
                List<String> neighbours = adjacencyList.getOrDefault(node, Collections.emptyList());
                // Push in reverse to preserve left-to-right order
                for (int i = neighbours.size() - 1; i >= 0; i--) {
                    if (!visited.contains(neighbours.get(i))) {
                        stack.push(neighbours.get(i));
                    }
                }
            }
        }
        return result;
    }

    // ── BFS ──────────────────────────────────────────────────────────────────

    /**
     * BFS using a queue.
     * Explores all neighbours at current depth before going deeper.
     * Returns nodes in visit order.
     */
    public List<String> bfs(String start) {
        List<String> result = new ArrayList<>();
        Set<String> visited = new HashSet<>();
        Queue<String> queue = new LinkedList<>();

        visited.add(start);
        queue.add(start);

        while (!queue.isEmpty()) {
            String node = queue.poll();
            result.add(node);
            for (String neighbour : adjacencyList.getOrDefault(node, Collections.emptyList())) {
                if (!visited.contains(neighbour)) {
                    visited.add(neighbour);
                    queue.add(neighbour);
                }
            }
        }
        return result;
    }

    /**
     * BFS shortest path between start and end.
     * Returns the path as a list of nodes, or an empty list if no path exists.
     */
    public List<String> bfsShortestPath(String start, String end) {
        if (start.equals(end)) return List.of(start);

        Set<String> visited = new HashSet<>();
        Queue<List<String>> queue = new LinkedList<>();
        queue.add(new ArrayList<>(List.of(start)));
        visited.add(start);

        while (!queue.isEmpty()) {
            List<String> path = queue.poll();
            String node = path.get(path.size() - 1);

            for (String neighbour : adjacencyList.getOrDefault(node, Collections.emptyList())) {
                if (!visited.contains(neighbour)) {
                    List<String> newPath = new ArrayList<>(path);
                    newPath.add(neighbour);
                    if (neighbour.equals(end)) return newPath;
                    visited.add(neighbour);
                    queue.add(newPath);
                }
            }
        }
        return Collections.emptyList(); // No path found
    }
}
